File: app/src/main/java/com/venditti/placares/fragments/PointsFragment.java
package com.venditti.placares.fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.ValueEventListener;
import com.venditti.placares.adapter.PointsAdapter;
import com.venditti.placares.config.ConfigFireBase;
import com.venditti.placares.databinding.FragmentPointsBinding;
import com.venditti.placares.model.Players;
import com.venditti.placares.model.Points;

import java.util.ArrayList;
import java.util.List;

public class PointsFragment extends Fragment {

    private List<Players> listPlayer;
    private PointsAdapter adapter;
    private FragmentPointsBinding binding;
    private int rodada = 1;
    private int maximoRodada;
    final String[] key = {""};
    public PointsFragment() {
        // Required empty public constructor
    }

    @Override
    public void onStart() {
        super.onStart();
        recoveryGame();
    }

    @Override
    public void onStop() {
        super.onStop();
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentPointsBinding.inflate(inflater);
        Bundle dados = getActivity().getIntent().getExtras();
        listPlayer = (ArrayList<Players>) dados.getSerializable("listaJogadores");

        configuraAdapter();

        binding.txtRodada.setText("Rodada = " + rodada);
        binding.btnCalcular.setOnClickListener(v -> atualizaPontos());

        getQtdRodada();
        return binding.getRoot();
    }

    private void configuraAdapter() {
        adapter = new PointsAdapter(listPlayer);
        //Configura RecyclerView
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(getActivity());
        binding.recyclerView.setLayoutManager(layoutManager);
        binding.recyclerView.setHasFixedSize(true);
        binding.recyclerView.setAdapter(adapter);
    }

    private void atualizaPontos() {
        if(rodada < maximoRodada) rodada ++; else rodada--;

        adapter.getListJogadores().forEach(p -> {
            int sum = p.getPoints().stream().mapToInt(Points::getSummation).sum();
            p.setTotal(sum);

            ConfigFireBase.getPlayerRef("Bisca", key[0])
                    .child(p.getName())
                    .setValue(p);
        });
        adapter.notifyDataSetChanged();
        binding.txtRodada.setText("Rodada = " + rodada);
    }

    private void recoveryGame() {

        ConfigFireBase.getDataGameBiscaReference()
                .addListenerForSingleValueEvent(new ValueEventListener() {
                    @Override
                    public void onDataChange(@NonNull DataSnapshot snapshot) {
                        for (DataSnapshot s : snapshot.getChildren()) {
                            key[0] = s.getKey();
                        }
                    }

                    @Override
                    public void onCancelled(@NonNull DatabaseError error) {
                    }
                });

    }

    private void getQtdRodada() {
        switch (listPlayer.size()) {
            case 3:
                maximoRodada = 17;
                break;
            case 4:
                maximoRodada = 12;
                break;
            case 5:
                maximoRodada = 10;
                break;
            case 6:
                maximoRodada = 8;
                break;
            case 7:
                maximoRodada = 7;
                break;
            case 8:
                maximoRodada = 6;
                break;
        }
    }
}

File: app/src/main/java/com/venditti/placares/fragments/PointsFragment.kt
package com.venditti.placares.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener
import com.venditti.placares.adapter.PointsAdapter
import com.venditti.placares.config.ConfigFireBase
import com.venditti.placares.databinding.FragmentPointsBinding
import com.venditti.placares.model.Players
import java.util.function.Consumer

class PointsFragment : Fragment() {
    private var listPlayer: List<Players?>? = null
    private var adapter: PointsAdapter? = null
    private var binding: FragmentPointsBinding? = null
    private var rodada = 1
    private var rodadasSubindo = true
    private var maximoRodada = 0
    val key = arrayOf<String?>("")
    override fun onStart() {
        super.onStart()
        recoveryGame()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        binding = FragmentPointsBinding.inflate(inflater)
        val dados = requireActivity().intent.extras
        listPlayer = dados!!.getSerializable("listaJogadores") as ArrayList<Players?>?

        configuraAdapter()
        binding!!.txtRodada.text = "Rodada = $rodada"
        binding!!.btnCalcular.setOnClickListener { v: View? -> atualizaPontos() }
        qtdRodada
        return binding!!.root
    }

    private fun configuraAdapter() {
        adapter = PointsAdapter(listPlayer, rodada.toString())
        //Configura RecyclerView
        val layoutManager: RecyclerView.LayoutManager = LinearLayoutManager(activity)
        binding!!.recyclerView.layoutManager = layoutManager
        binding!!.recyclerView.setHasFixedSize(false)
        binding!!.recyclerView.adapter = adapter
    }

    private fun atualizaPontos() {
        if (rodadasSubindo) {
            rodada++
            adapter!!.setRodada(rodada.toString())
            binding!!.txtRodada.text = "Rodada = $rodada"
        } else {
            rodada--
            binding!!.txtRodada.text = "Rodada = " + rodada + "v"
            adapter!!.setRodada(rodada.toString() + "v")
        }
        if (rodada > maximoRodada) {
            rodadasSubindo = false
        } else if (rodada < 1) binding!!.txtRodada.text = "FIM DE JOGO"
        adapter?.listJogadores?.forEach(Consumer { p: Players? ->
            p?.total = p?.points?.values?.sumOf { it?.summation ?: 0 }!!

            ConfigFireBase.getPlayerRef("Bisca", key[0])
                    .child(p.name)
                    .setValue(p)
        })
        adapter!!.notifyDataSetChanged()
    }

    private fun recoveryGame() {
        ConfigFireBase.dataGameBiscaReference
                .addListenerForSingleValueEvent(object : ValueEventListener {
                    override fun onDataChange(snapshot: DataSnapshot) {
                        for (s in snapshot.children) {
                            key[0] = s.key
                        }
                    }

                    override fun onCancelled(error: DatabaseError) {}
                })
    }

    private val qtdRodada: Unit
        get() {
            when (listPlayer!!.size) {
                3 -> maximoRodada = 17
                4 -> maximoRodada = 12
                5 -> maximoRodada = 10
                6 -> maximoRodada = 8
                7 -> maximoRodada = 7
                8 -> maximoRodada = 6
            }
        }
}